### PR TITLE
S4 incorporation

### PR DESCRIPTION
**Pull request by Rozhkin-Aleksandr**
This pull request introduces the integration of the S4, S4D, and DSS State Space Models into our framework, resolving issue #13. The implementation ensures seamless compatibility with our existing training and evaluation workflows.

**Key Changes & Features:**

*   **Core Model Implementation:**
    *   A new `s4.py` file provides a comprehensive realization of the S4 block, incorporating all necessary modifications.
    *   The `modeling_s4.py` file defines the core model classes for the sequence classification and pretraining.
*   **New Tasks & Configurations:**
    *   `tasks.py` has been updated to include new tasks for S4 model experiments in sequence classification and Masked Language Modeling (MLM) pretraining.
    *   New configurations for S4, S4D, and DSS models have been written for Pathfinder-32 and C4 pretraining (supporting bidirectional SSMs as requested).
*   **Experiments:**
    *   The newly integrated models have been thoroughly tested with the provided configurations to confirm correct functionality and adherence to framework standards.

### CODE REVIEW

File: configs/bert/deepspeed_config_seq128.json [hunk 1]
```diff
@@ -21,7 +21,7 @@
   "wall_clock_breakdown": false,
 
   "fp16": {
-    "enabled": true,
+    "enabled": false,
```
Comment (andrewargatkiny) [inline configs/bert/deepspeed_config_seq128.json:24]: Please revert to the previous value.

File: src/other_models/s4.py [hunk 2]
```diff
@@ -0,0 +1,1576 @@
+"""Standalone version of Structured (Sequence) State Space (S4) model."""
```
Comment (andrewargatkiny) [inline src/other_models/s4.py:1]: Please reference the original code (and/ or papers) upon which your code is based, and its license, if applicable.

File: src/other_models/s4.py [hunk 3]
```diff
@@ -0,0 +1,1576 @@
...
+import math
+import numpy as np
+from scipy import special as ss
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from einops import rearrange, repeat
+import opt_einsum as oe
+
+contract = oe.contract
```
Comment (andrewargatkiny) [inline src/other_models/s4.py:14]: Why did you decide to use `opt_einsum` instead of `torch.einsum` like in the original code? Did you measure that it's faster?

Also, [torch.einsum](https://docs.pytorch.org/docs/stable/generated/torch.einsum.html) uses opt_einsum if its installed, so I suggest to proceed with the original code if it's not slower.

File: src/other_models/s4.py [hunk 4]
```diff
@@ -0,0 +1,1576 @@
...
+    """Initializes multi-GPU-friendly python logger."""
+
+    logger = logging.getLogger(name)
+    logger.setLevel(level)
+
+    # this ensures all logging levels get marked with the rank zero decorator
+    # otherwise logs would get multiplied for each GPU process in multi-GPU setup
+ 
+    return logger
+log = get_logger(__name__)
```
Comment (andrewargatkiny) [inline src/other_models/s4.py:28]: We already have a logger with correct multi-GPU handling in the project. Please consider using it instead.

File: src/other_models/modeling_s4.py [hunk 5]
```diff
@@ -0,0 +1,592 @@
...
+        For non-embedding count (default), the position embeddings get subtracted.
+        The token embeddings would too, except due to the parameter sharing these
+        params are actually used as weights in the final layer, so we include them.
+        """
+        n_params = sum(p.numel() for p in self.parameters())
+        if non_embedding:
+            n_params -= self.embeddings.word_embeddings.weight.numel()
+        return n_params
+
+    def init_bert_weights(self, module):
```
Comment (andrewargatkiny) [inline src/other_models/modeling_s4.py:285]: Please check that this function affects none of the parameters of the `S4 layer` module, because it has very intricate initialization having a key impact on its performance.

Comment (andrewargatkiny) [reply src/other_models/modeling_s4.py:285]: Update: found at least one such case.

File: src/other_models/s4.py (hunk 3 above)
Comment (andrewargatkiny) [reply src/other_models/s4.py:14]: Update: my previous comment is not relevant since this code copies another implementation of S4, not the one I thought of.

File: src/other_models/s4.py [hunk 6]
```diff
@@ -0,0 +1,1576 @@
...
+        *[ssm(measure, N, R, S // len(measures), **ssm_args) for measure in measures]
+    )
+    w = torch.cat(w, dim=0) # (S N)
+    P = torch.cat(P, dim=1) # (R S N)
+    B = torch.cat(B, dim=0) # (S N)
+    V = torch.cat(V, dim=0) # (S N N)
+    return w, P, B, V
+
+
+class OptimModule(nn.Module):
```
Comment (andrewargatkiny) [inline src/other_models/s4.py:535]: This functionally wouldn't correctly work with out framework. And not only this one, but many architectures and models would benefit from granular learning rates and weight decay values per module/parameter.

It needs a thoughtful consideration on how to properly implement it. I suggest that a config for a particular layer type should contain a dict with custom lr/ wd values for some of its parameters.

Please implement this feature in a standalone PR.

File: src/other_models/s4.py [hunk 7]
```diff
@@ -0,0 +1,1576 @@
...
+
+        See the class SSKernel for the kernel constructor which accepts kernel_args. Relevant options that are worth considering and tuning include "mode" + "measure", "dt_min", "dt_max", "lr"
+
+        Other options are all experimental and should not need to be configured
+        """
+
+        super().__init__()
+        self.d_model = config.d_model
+        self.d_state = config.d_state
+        self.num_hidden_layers = config.num_hidden_layers
```
Comment (andrewargatkiny) [inline src/other_models/s4.py:1388]: I believe this parameter is redundant here. Are there any other unused params in this initialization?

File: src/other_models/s4.py (hunk 2 above)
Comment (andrewargatkiny) [reply src/other_models/s4.py:1]: Also cite this useful manual which details hyperparams for this implementation:
https://github.com/IdoAmos/not-from-scratch/tree/main/src/models/s4

File: src/other_models/modeling_s4.py
Comment (andrewargatkiny) [inline src/other_models/modeling_s4.py:1]: This code is pretty much repeating `src/other_models/modeling.py`. As well as the new tasks in `utils/tasks.py`. This underlines a deeper problem: whether each new sequence-mixer/ architecture layer type should have its own outer model layers for different tasks. I'm proposing to implement a single interface which could accommodate many different sequence-mixers – see #35. 

That way, incorporating a new sequence-mixer architecture into our framework would be as simple as adding its source code file(s) and writing a few lines of harness to register the new layer type and its config type.

For now, please postpone working on this PR until #35 is finished, and then use that new feature to incorporate S4 layer into the framework.

**Review by andrewargatkiny — CHANGES_REQUESTED** [review]
